### PR TITLE
Add new `Style/ArrayIntersect` cop

### PR DESCRIPTION
**Pull request by KirIgor**
In ruby 3.1 `Array#intersect?` has been [added](https://github.com/ruby/ruby/pull/1972) which is faster than `(array1 & array2).any?` since the new method avoids creating an intermediate array while evaluating for common elements. Also new method returns true as soon as it finds a common element between arrays. 

Also new method is more readable.

I have added cop `Style/ArrayIntersect` which identifies and corrects this cases. It can find `&` usage with `.any?`, `.present?`, `.empty?` and `.blank?` methods.

Moved from [rubocop-perfomance#312](https://github.com/rubocop/rubocop-performance/pull/312).

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `new_style_intersect_predicate_cop.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by KirIgor** [comment]
Speedup confirm by @koic (copied from [rubocop-perfomance#312](https://github.com/rubocop/rubocop-performance/pull/312#issuecomment-1334802641).

```ruby
require 'bundler/inline'

gemfile(true) do
  gem 'benchmark-ips'
end

Benchmark.ips do |x|
  x.report('&') { ([1, 2, 3] & [2, 3, 4]).any? }
  x.report('intersect?') { [1, 2, 3].intersect?([2, 3, 4]) }

  x.compare!
end
```

```
% ruby intersect.rb
(snip)

Warming up --------------------------------------
                   &   452.504k i/100ms
          intersect?   708.976k i/100ms
Calculating -------------------------------------
                   &      4.458M (± 1.8%) i/s -     22.625M in   5.076782s
          intersect?      6.818M (± 4.8%) i/s -     34.031M in   5.004272s

Comparison:
          intersect?:  6817812.7 i/s
                   &:  4457969.2 i/s - 1.53x  (± 0.00) slower
```

**Comment by bbatsov** [comment]
I like the cop in general, but I'm wondering about the name - perhaps it should be `ArrayIntersect`, so it's a bit clearer what is it about. Anyways, I'm open to suggestions - naming continues to be hard. :-)

**Comment by KirIgor** [comment]
Naming is definitely hard. :slightly_smiling_face: At first it was abstract `IntersectionCheck`. I think `ArrayIntersect` is good.

**Comment by KirIgor** [comment]
Name was changed to `Style/ArrayIntersect`.

**Comment by koic** [comment]
Yeah, I also think `ArrayIntersect` is a better name!

File: lib/rubocop/cop/style/array_intersect.rb [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+module RuboCop
+  module Cop
+    module Style
+      # In Ruby 3.1, `Array#intersect?` has been added.
+      #
+      # This cop identifies places where `(receiver & argument).any?`
+      # can be replaced by `receiver.intersect?(argument)`.
+      #
+      # The `receiver.intersect?(argument)` method is faster than
+      # `(receiver & argument).any?` and is more readable.
```
Comment (koic) [inline lib/rubocop/cop/style/array_intersect.rb:12]: A comment about `AllCops:ActiveSupportExtensionsEnabled` would help users understand. e.g.:

```ruby
# `(receiver & argument).blank?` and `(receiver & argument).present?` are detected
# if `AllCops:ActiveSupportExtensionsEnabled` is `true`.
```

File: lib/rubocop/cop/style/array_intersect.rb [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      #
+      #   # good
+      #   receiver.intersect?(argument)
+      #
+      #   # bad
+      #   (receiver & argument).blank?
+      #   (receiver & argument).empty?
+      #
+      #   # good
+      #   !receiver.intersect?(argument)
```
[the remark this answers is not in the document]
Comment (KirIgor) [reply lib/rubocop/cop/style/array_intersect.rb:27]: Thank you!

File: lib/rubocop/cop/style/array_intersect.rb (hunk 1 above)
Comment (KirIgor) [reply lib/rubocop/cop/style/array_intersect.rb:12]: Fixed in discussion below.

**Comment by bbatsov** [comment]
Two small notes:

- Add some `@safety` section to the docs, as clearly that cop is unsafe as we can't be sure if something is an array or not.
- Perhaps using in the examples names like `array1` and `array2` will make it clearer that this is array specific functionality.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**